File: 1043. Partition Array for Maximum Sum/maxSumAfterPartitioning.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        int maxSumAfterPartitioning(vector<int>& arr, int k) {
            int len = arr.size();
    
            vector<int> dp(len+1 , 0);
    
            for(int i=1 ; i<=len; i++){
                int maximum = 0;
                int maxSum = 0;
    
                for(int j=1; j<=k && i-j>=0; j++){
    
                    maximum = max(maximum, arr[i-j]);
                    maxSum = max(maxSum, dp[i-j] + maximum * j);
                }
    
                dp[i] = maxSum;
            }
    
            return dp[len];
        }
    };

File: 1043. Partition Array for Maximum Sum/maxSumAfterPartitioning.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var maxSumAfterPartitioning = function (nums, k) {
    let len = nums.length

    let dp = new Array(len + 1).fill(0)

    for (let i = 1; i <= len; i++) {
        let max = 0;
        let maxSum = 0;

        for (let j = 1; j <= k && i - j >= 0; j++) {
            max = Math.max(max, nums[i - j])
            //add the current partition sum, which is max * j 
            maxSum = Math.max(maxSum, dp[i - j] + max * j)
        }
        dp[i] = maxSum;

    }
    return dp[len]

};